tls/extensions/s2n_client_key_share.c

```diff
@@ -342,80 +323,85 @@ static int s2n_client_key_share_recv_pq_hybrid(struct s2n_connection *conn, stru
+ * group that we already have a key share for, since retries are expensive.
+ *
+ * This method modifies our group selection based on what keyshares are available.
+ * It then stores the client keyshare for the selected group, or initiates a retry
+ * if no valid keyshares are available.
+ */
+static int s2n_client_key_share_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
     POSIX_ENSURE_REF(conn);
     POSIX_ENSURE_REF(extension);
 
-    const struct s2n_ecc_preferences *ecc_pref = NULL;
-    POSIX_GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
-    POSIX_ENSURE_REF(ecc_pref);
-
-    const struct s2n_kem_preferences *kem_pref = NULL;
-    POSIX_GUARD(s2n_connection_get_kem_preferences(conn, &kem_pref));
-    POSIX_ENSURE_REF(kem_pref);
-
     uint16_t key_shares_size;
     POSIX_GUARD(s2n_stuffer_read_uint16(extension, &key_shares_size));
-    POSIX_ENSURE(s2n_stuffer_data_available(extension) >= key_shares_size, S2N_ERR_BAD_MESSAGE);
+    POSIX_ENSURE(s2n_stuffer_data_available(extension) == key_shares_size, S2N_ERR_BAD_MESSAGE);
 
-    uint16_t named_group, share_size;
-    bool match_found = false;
-    /* bytes_processed is declared as a uint32_t to avoid integer overflow in later calculations */
-    uint32_t bytes_processed = 0;
+    uint16_t named_group = 0, share_size = 0;
+    struct s2n_blob key_share_blob = { 0 };
+    struct s2n_stuffer key_share = { 0 };
 
-    while (bytes_processed < key_shares_size) {
+    uint16_t keyshare_count = 0;
+    while(s2n_stuffer_data_available(extension) > 0) {
         POSIX_GUARD(s2n_stuffer_read_uint16(extension, &named_group));
         POSIX_GUARD(s2n_stuffer_read_uint16(extension, &share_size));
-
         POSIX_ENSURE(s2n_stuffer_data_available(extension) >= share_size, S2N_ERR_BAD_MESSAGE);
-        bytes_processed += share_size + S2N_SIZE_OF_NAMED_GROUP + S2N_SIZE_OF_KEY_SHARE_SIZE;
 
-        struct s2n_blob key_share_blob = { .size = share_size, .data = s2n_stuffer_raw_read(extension, share_size) };
-        POSIX_ENSURE_REF(key_share_blob.data);
-        struct s2n_stuffer key_share = { 0 };
+        POSIX_GUARD(s2n_blob_init(&key_share_blob,
+            s2n_stuffer_raw_read(extension, share_size), share_size));
         POSIX_GUARD(s2n_stuffer_init(&key_share, &key_share_blob));
         POSIX_GUARD(s2n_stuffer_skip_write(&key_share, share_size));
+        keyshare_count++;
 
         /* Try to parse the share as ECC, then as PQ/hybrid; will ignore
          * shares for unrecognized groups. */
-        POSIX_GUARD(s2n_client_key_share_recv_ecc(conn, &key_share, named_group, &match_found));
-        POSIX_GUARD(s2n_client_key_share_recv_pq_hybrid(conn, &key_share, named_group, &match_found));
+        POSIX_GUARD(s2n_client_key_share_recv_ecc(conn, &key_share, named_group));
+        POSIX_GUARD(s2n_client_key_share_recv_pq_hybrid(conn, &key_share, named_group));
     }
 
+    /* During a retry, the client should only have sent one keyshare */
```

Comment:
Aren't we only ever sending a single keyshare?

Reply:
We are, but this is server-side code to process the shares a client sent, so there's no guarantee we're talking to an S2N client.